### ISSUE TITLE
Deadlock in MultiProc with resource profiler inside docker

### ISSUE THREAD

**Issue by oesteban** [nipy/nipype#1994]
### Summary
We are experiencing this with mriqc on CircleCI. The workflow runs for a couple of minutes and gets locked. Using Linear works well.

### How to replicate the behavior
https://circleci.com/gh/oesteban/mriqc/887

Can't replicate in a different machine with bare-metal installation.

### Script/Workflow details
This branch https://github.com/oesteban/mriqc/tree/enh/ReduceSpatialNormalizationRuntime

Potentially related to the number of threads I'm setting in an ants.Registration interface interacting with the number of processors of multiproc.

### Platform details:
ad hoc docker image

**Comment by ccraddock** [comment]
Does the same problem occur when the resource profiler is turned off?

**Comment by chrisfilo** [comment]
You are requesting 4 cores for a single job while running the whole workflow with 2 cores enabled. The job will never run since MultiProc does not have 4 cores available. This is expected behaviour.

Maybe we should give a better error message in such cases.

**Comment by oesteban** [comment]
@ccraddock Nope, I was exactly trying that. When I don't hook up the callback it doesn't seem to get locked

**Comment by oesteban** [comment]
@chrisfilo you refer to the node, or MultiProc is checking the number of available cores regardless what I set in the plugin_args?

**Comment by chrisfilo** [comment]
If I understand your code correctly: the ants Node requests 4 cores, but you tell MultiProc plugin there are only 2 cores available. AFAIK if number of available cores is passed to MultiProc it will not try to estimate this.

**Comment by effigies** [comment]
This line is what's going to kill you no matter what, if `n_procs < 4`:

https://github.com/oesteban/mriqc/blob/enh/ReduceSpatialNormalizationRuntime/mriqc/bin/mriqc_run.py#L211

**Comment by oesteban** [comment]
@chrisfilo Ok, so we are talking at the node level :+1: . Then, I'm trying to trick MultiProc: while I set the `num_threads` input of the ants interface to 4, I am also setting the `num_threads` of the Node to 2, with the hope that MultiProc will check that one and not the other.

@effigies shouldn't, because ants_threads is set only for the interface and n_procs for the node.

**Comment by oesteban** [comment]
@ccraddock scratch my previous comment, it also happens without hooking the callback.

**Comment by ccraddock** [comment]
In line at security so can't look at code at the moment. If you request more threads or mem than is available on the system it should throw an exception that kills the execution of the entire pipeline. If it doesn't do this correctly, then it could result in an unresolvable dependency that prevents pipeline from finishing. Maybe this is what your seeing?

**Comment by oesteban** [comment]
Yes, we are looking at a mixup. To provide more info I logged into a Circle build through ssh. This is what we see from within the docker image:

```
(venv-system) ubuntu@box465:~$ docker run -ti  --entrypoint=python poldracklab/mriqc:latest -c 'import multiprocessing; print(multiprocessing.cpu_count())'
32
(venv-system) ubuntu@box465:~$ docker run -ti  --entrypoint=python poldracklab/mriqc:latest -c 'import psutil; print(psutil.virtual_memory())'
svmem(total=257948532736, available=232078548992, percent=10.0, used=25198755840, free=118734651392, active=55581032448, inactive=64164683776, buffers=84791296, cached=113930334208, shared=325550080)
```
I'll check I'm not requesting more memory than available.

Regarding the resources: it is likely the resource limitations are not working completely right. I have the impression that the workflows aren't killed when you request more cores or more memory than available. I'll try to gather more info on this.

When revising this, we could add a nipype config setting like `multiproc_resource_limit_error`, default to `True`. If `False`, a warning would be issued and the resources adapted to the availability.

**Comment by ccraddock** [comment]
Adapting the resources will likely require a lot more and is particularly problematic for memory. Few tools have the ability to adapt their memory usage and going over the limit will result in either bringing the system to a halt or the process being killed by the out of memory killer.
